source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -64,13 +133,39 @@ bool LocalRateLimiterImpl::requestAllowed() const {
     synchronizer_.syncPoint("allowed_pre_cas");
 
     // Loop while the weak CAS fails trying to subtract 1 from expected.
-  } while (!tokens_.compare_exchange_weak(expected_tokens, expected_tokens - 1,
-                                          std::memory_order_relaxed));
+  } while (!tokens.tokens.compare_exchange_weak(expected_tokens, expected_tokens - 1,
+                                                std::memory_order_relaxed));
 
   // We successfully decremented the counter by 1.
   return true;
 }
 
+bool LocalRateLimiterImpl::requestAllowed() const { return requestAllowedHelper(tokens_); }
+
+bool LocalRateLimiterImpl::requestAllowed(
+    const Envoy::RateLimit::LocalDescriptor& descriptor) const {
+  auto it = tokens_per_descriptor_.find(descriptor);
+  if (it == tokens_per_descriptor_.end()) {
+    return false;
+  }
+  return requestAllowedHelper(it->second->token);
+}
```

Comment:
Can you just collapse this into a single function? It's not clear to me why we have to have separate functions here, + another `findDescriptor` function. Can you just pass in all the state needed to do the requestAllowed() computation and have this object sort it out?

Reply:
so combine requestAllowed with findDescriptor ? as we need requestAllowedHelper so that requestAllowed and requestAllowed for descriptor can share same synchronize code

Reply:
The only public function should be `requestAllowed(...)`. Pass in any data needed to compute this, even if it's an empty list of descriptors. Then the function can do what it needs to do and should still be efficient. This will greatly simplify the overall logic.